src/sqlfluff/core/parser/grammar/greedy.py

```diff
@@ -118,7 +118,7 @@ def greedy_match(
                     # a requirement, then we can't use it. Carry on...
                     if not allowable_match:
                         # Update our buffers and continue onward
-                        seg_bank = pre + mat.matched_segments
+                        seg_bank = seg_bank + pre + mat.matched_segments
```

Comment:
This is the bug that prevented parsing working. Think this might have caused a few parsing errors for `StartsWith`.